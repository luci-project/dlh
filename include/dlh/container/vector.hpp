/*! \file
 *  \brief vector class
 */
#pragma once

#include <dlh/alloc.hpp>
#include <dlh/assert.hpp>
#include <dlh/types.hpp>
#include <dlh/utility.hpp>
#include <dlh/container/optional.hpp>
#include <dlh/stream/buffer.hpp>

/*! \brief Vector class influenced by standard [template/cxx] library
 */
template<class T> class Vector {
	/*! \brief Array with entries
	 * dynamically allocated / increased
	 */
	T* _element;

	/*! \brief Number of entries
	 */
	size_t _size;

	/*! \brief Current maximum capacity
	 */
	size_t _capacity;

	/*! \brief Expand capacity
	 */
	inline void expand() {
		reserve(0 < _capacity ? 2 * _capacity : 16);
	}

 public:
	Vector(const Vector&) = default;
	Vector(Vector&&) = default;

	/*! \brief Constructor
	 */
	Vector() : _element(nullptr), _size(0), _capacity(0) {}

	/*! \brief Constructor with initial _capacity
	 * \param capacity number of initial values
	 * \param init initial values
	 * \param invalid object for invalid returns
	 */
	explicit Vector(size_t capacity) : _element(nullptr), _size(capacity), _capacity(capacity) {
		if (capacity > 0) {
			_element = reinterpret_cast<T*>(malloc(capacity * sizeof(T)));
			assert(_element != nullptr);
		}
	}

	/*! \brief Range constructor
	 * \param first First element in range
	 * \param last Last element in range
	 * \param invalid object for invalid returns
	 */
	template<typename I>
	explicit Vector(I first, I last) : Vector(0) {
		for (I i = first; i < last; ++i)
			emplace_back(*i);
	}

	/*! \brief Copy constructor
	 * \param other other vector
	 */
	template<typename I>
	explicit Vector(const Vector & other) : Vector(other.begin(), other.end()) {}


	/*! \brief Destructor
	 */
	~Vector() {
		resize(0);
		free(_element);
	}

	/*! \brief access specified element
	 * \param pos position of the element to return
	 * \return element or T() if out of bounds
	 * \note This function differs to the standard library due to the lack of exceptions
	 */
	inline Optional<T> at(int i) {
		if (i < _size)
			return { _element[i] };
		else
			return {};
	}

	/*! \brief Access the first element
	 * \return first element
	 */
	inline T& front() {
		return operator[](0);
	}

	/*! \brief Access the last element
	 * \return last element
	 */
	inline T& back() {
		return operator[](_size - 1);
	}

	/*! \brief direct access to the underlying array
	 * \return Pointer to data
	 */
	inline T* data() {
		return _element;
	}

	/*! \brief Checks if the container has no element
	 */
	inline bool empty() const {
		return _size == 0;
	}

	/*! \brief Get the number of elements
	 * \return number of entires
	 */
	inline size_t size() const {
		return _size;
	}

	/*! \brief Get the number of elements that can be held in currently allocated storage
	 * \return current capacity
	 */
	inline size_t capacity() const {
		return _capacity;
	}

	/*! \brief Erases all elements from the container
	 */
	inline void clear() {
		_size = 0;
	}

	/*! \brief Increase the capacity of the vector
	 * \param capacity new capacity
	 */
	inline void reserve(size_t capacity) {
		if (capacity > this->_capacity) {
			_element = reinterpret_cast<T*>(realloc(reinterpret_cast<void *>(_element), capacity * sizeof(T)));
			this->_capacity = capacity;
		}
	}

	/*! \brief Resizes the container
	 * \param count new size of the container
	 */
	inline void resize(size_t count) {
		if (count < _size) {
			// Free items (
			for (size_t i = count; i < _size; ++i)
				_element[i].~T();
			_size = count;
		}
		reserve(count);
	}

	/*! \brief Creates an element at the end
	 * \param args Arguments to create the value
	 */
	template<typename... ARGS>
	inline void emplace_back(ARGS&&... args) {
		if (_capacity == _size)
			expand();

		new (_element + _size) T(forward<ARGS>(args)...);
		_size++;
	}

	/*! \brief Adds an element to the end
	 * \param value the value of the element to append
	 */
	inline void push_back(const T& value) {
		emplace_back(value);
	}

	/*! \brief Adds an element to the end
	 * \param value the value of the element to append
	 */
	inline void push_back(T&& value) {
		emplace_back(forward<T>(value));
	}

	/*! \brief Creats element at the specified location
	 * \param pos position
	 * \param args Arguments to create the value
	 */
	template<typename... ARGS>
	inline void emplace(size_t pos, ARGS&&... args) {
		if (pos == _capacity)
			expand();

		if (pos <= _size) {
			for (size_t i = _size; i > pos; --i)
				new (_element + i) T(move(_element[i - 1]));

			new (_element + pos) T(forward<ARGS>(args)...);
			_size++;
		}
	}

	/*! \brief Inserts element at the specified location
	 * \param pos position
	 * \param value value
	 */
	inline void insert(size_t pos, const T& value) {
		emplace(pos, value);
	}

	/*! \brief Inserts element at the specified location
	 * \param pos position
	 * \param value value
	 */
	inline void insert(size_t pos, T&& value) {
		emplace(pos, forward<T>(value));
	}

	/*! \brief Remove the last element
	 * \return last element
	 */
	inline Optional<T> pop_back() {
		return 0 == _size ? Optional<T>() : Optional<T>(_element[--_size]);
	}

	/*! \brief Remove element at the specified location
	 * \param pos position
	 * \return value (if available)
	 */
	inline Optional<T> remove(size_t pos) {
		if (pos < _size) {
			auto r = Optional<T>(_element[pos]);
			for (size_t i = pos; i < _size - 1; i++)
				new (_element + i) T(move(_element[i - 1]));

			_size--;
			return r;
		} else {
			return Optional<T>();
		}
	}

	/*! \brief Access Element
	 * \note undefined behaviour if index is out of bounds
	 * \param i index
	 */
	inline T & operator[](int i) {
		assert(i >= 0 && static_cast<size_t>(i) < _size);
		return _element[i];
	}

	/*! \brief Copy Assignment
	 * \param other
	 */
	Vector<T>& operator=(const Vector<T>& other) {
		// No self assignment
		if (this != &other) {
			resize(0);
			for (auto & e : other)
				emplace_back(e);
		}
		return *this;
	}

	/*! \brief Move Assignment
	 * \param other
	 */
	Vector<T>& operator=(Vector<T>&& other) = default;

	/*! \brief Concatenate vector
	 * \param other vector
	 * \return reference to vector
	 */
	Vector<T> & operator+=(const Vector<T>& other) {
		reserve(_size + other._size);
		for (size_t i = 0; i < other._size; ++i)
			push_back(i);

		return *this;
	}

	/*! \brief Concatenate element
	 * \param element element
	 * \return reference to vector
	 */
	Vector<T> & operator+=(const T& element) {
		push_back(element);
		return *this;
	}

	/*! \brief Concatenate two vectors
	 * \param other vector
	 * \return new vector with all elements of this and other vector
	 */
	Vector<T> operator+(const Vector<T>& other) const {
		Vector<T> r = *this;
		r += other;
		return r;
	}

	/*! \brief Concatenate element
	 * \param element element to Concatenate
	 * \return new vector with all elements of this vector and element
	 */
	Vector<T> operator+(const T& element) const {
		Vector<T> r = *this;
		r += element;
		return r;
	}

	/*! \brief Vector iterator
	 */
	class Iterator {
		T* i;

	 public:  //NOLINT
		explicit Iterator(T* p) : i(p) {}

		Iterator& operator++() {
			i++;
			return *this;
		}

		Iterator& operator--() {
			i--;
			return *this;
		}

		T& operator*() {
			return *i;
		}

		bool operator==(const Vector<T>::Iterator& other) const {
			return i == other.i;
		}

		bool operator!=(const Vector<T>::Iterator& other) const {
			return i != other.i;
		}
	};

	inline Vector<T>::Iterator begin() const {
		return Vector<T>::Iterator(_element);
	}

	inline Vector<T>::Iterator end() const {
		return Vector<T>::Iterator(_element + _size);
	}
};


/*! \brief Print contents of a vector
 *
 *  \param bs Target Bufferstream
 *  \param val Vector to be printed
 *  \return Reference to BufferStream os; allows operator chaining.
 */
template<typename T>
static inline BufferStream& operator<<(BufferStream& bs, Vector<T> & val) {
	bs << "{ ";
	bool p = false;
	for (const auto & v : val) {
		if (p)
			bs << ", ";
		p = true;
		bs << v;
	}
	return bs << '}';
}
