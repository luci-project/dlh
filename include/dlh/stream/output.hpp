#pragma once

#include <dlh/syscall.hpp>
#include <dlh/stream/string.hpp>

template<size_t BUFFERSZ>
class OutputStream : public StringStream<BUFFERSZ> {
 protected:
	int fd;

 public:
	/*! \brief Default constructor  */
	explicit OutputStream(int fd) : fd(fd) {}

	/*! \brief Destructor */
	virtual ~OutputStream() {}

	/*! \brief Flush the buffer.
	 */
	virtual void flush() override {
		size_t l = 0;
		while (l < this->_pos) {
			auto r = Syscall::write(fd, this->_bufptr + l, this->_pos - l);
			if (!r.valid())
				break;
			else
				l += r.value();
		}
		this->_pos = 0;
	}

};

extern OutputStream<1024> cout, cerr;
