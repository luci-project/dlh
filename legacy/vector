#pragma once

#include <dlh/container/vector.hpp>

namespace std {

template<typename T>
using vector = Vector<T>;

} // namespace std
