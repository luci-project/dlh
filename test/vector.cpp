#include <dlh/stream/output.hpp>
#include <dlh/container/vector.hpp>

#include "foo.hpp"

int main(int argc, const char *argv[]) {
	(void) argc;
	(void) argv;

	const Foo five(5);

	// Test 1
	{
		Vector<Foo> s;
		auto p = s.push_front(13);
		auto q = s.insert(p, 666);
		s.push_back(Foo(3));
		s.push_back(five);
		s.emplace_front(42);
		s.emplace(q, Foo(23));
		auto x = s.insert(q, 7);
		s.erase(x);

		for (const auto & i : s)
			cout << " - Foo #" << i.uid << " = " << i.id << endl;
		cout << endl;
	}

	// Test 2
	{
		Vector<Foo> bla = { 1001, 1002 };
		bla.push_back(1003);
		bla.push_front(1000);
		bla.emplace_back(1004);
		{
			const Foo thousandandfive(five.id + 1000);
			bla += thousandandfive;
		}
		for (const auto & i : bla)
			cout << " - Foo #" << i.uid << " = " << i.id << endl;
		cout << endl;
	}

	return 0;
}
