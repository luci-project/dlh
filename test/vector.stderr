Foo #0 (5) created
Foo #1 (13) created
Foo #2 (13) created (move from Foo #1)
Foo #1 (13) deleted
Foo #3 (666) created
Foo #4 (666) created (move from Foo #3)
Foo #3 (666) deleted
Foo #5 (3) created
Foo #6 (3) created (move from Foo #5)
Foo #5 (3) deleted
Foo #7 (5) created (copy from Foo #0)
Foo #8 (5) created (move from Foo #7)
Foo #7 (5) deleted
Foo #9 (3) created (move from Foo #6)
Foo #6 (3) deleted
Foo #10 (666) created (move from Foo #4)
Foo #4 (666) deleted
Foo #11 (13) created (move from Foo #2)
Foo #2 (13) deleted
Foo #12 (42) created
Foo #13 (23) created
Foo #14 (5) created (move from Foo #8)
Foo #8 (5) deleted
Foo #15 (3) created (move from Foo #9)
Foo #9 (3) deleted
Foo #16 (666) created (move from Foo #10)
Foo #10 (666) deleted
Foo #17 (13) created (move from Foo #11)
Foo #11 (13) deleted
Foo #18 (23) created (move from Foo #13)
Foo #13 (23) deleted
Foo #19 (7) created
Foo #20 (5) created (move from Foo #14)
Foo #14 (5) deleted
Foo #21 (3) created (move from Foo #15)
Foo #15 (3) deleted
Foo #22 (666) created (move from Foo #16)
Foo #16 (666) deleted
Foo #23 (13) created (move from Foo #17)
Foo #17 (13) deleted
Foo #24 (23) created (move from Foo #18)
Foo #18 (23) deleted
Foo #25 (7) created (move from Foo #19)
Foo #19 (7) deleted
Foo #25 (7) deleted
Foo #26 (23) created (move from Foo #24)
Foo #24 (23) deleted
Foo #27 (13) created (move from Foo #23)
Foo #23 (13) deleted
Foo #28 (666) created (move from Foo #22)
Foo #22 (666) deleted
Foo #29 (3) created (move from Foo #21)
Foo #21 (3) deleted
Foo #30 (5) created (move from Foo #20)
Foo #20 (5) deleted
Foo #12 (42) deleted
Foo #26 (23) deleted
Foo #27 (13) deleted
Foo #28 (666) deleted
Foo #29 (3) deleted
Foo #30 (5) deleted
Foo #31 (1001) created
Foo #32 (1002) created
Foo #33 (1003) created
Foo #34 (1003) created (move from Foo #33)
Foo #33 (1003) deleted
Foo #35 (1000) created
Foo #36 (1003) created (move from Foo #34)
Foo #34 (1003) deleted
Foo #37 (1002) created (move from Foo #32)
Foo #32 (1002) deleted
Foo #38 (1001) created (move from Foo #31)
Foo #31 (1001) deleted
Foo #39 (1000) created (move from Foo #35)
Foo #35 (1000) deleted
Foo #40 (1004) created
Foo #41 (1005) created
Foo #42 (1005) created (copy from Foo #41)
Foo #41 (1005) deleted
Foo #39 (1000) deleted
Foo #38 (1001) deleted
Foo #37 (1002) deleted
Foo #36 (1003) deleted
Foo #40 (1004) deleted
Foo #42 (1005) deleted
Foo #0 (5) deleted
